lib/puppet/provider/service/base.rb

```diff
@@ -30,17 +30,23 @@ def getps
   # parameter.
   def getpid
     @resource.fail "Either stop/status commands or a pattern must be specified" unless @resource[:pattern]
-    regex = Regexp.new(@resource[:pattern])
     ps = getps
+
     self.debug "Executing '#{ps}'"
-    IO.popen(ps) { |table|
-      table.each_line { |line|
-        if regex.match(line)
-          self.debug "Process matched: #{line}"
-          ary = line.sub(/^\s+/, '').split(/\s+/)
-          return ary[1]
-        end
-      }
+    table = Puppet::Util::Execution.execute(ps)
+
+    # Processes can add arbitrary sequences of bytes to the cmdline retrieved by
+    # 'ps -f', so drop encodings and match as binary.
+    regex = @resource[:pattern].dup
+    regex = Regexp.new(regex.force_encoding("ASCII-8BIT"))
+    table.force_encoding("ASCII-8BIT")
+
+    table.each_line { |line|
+      if regex.match(line)
+        self.debug "Process matched: #{line}"
+        ary = line.sub(/^\s+/, '').split(/\s+/)
```

Comment:
For `ASCII-8BIT` regex matches outside the ASCII range are done on a byte-by-byte basis. The test case in this patchset exercises this by matching the pattern `"majestic m\xd8\xd8se"` against a Latin-1 encoded file.

For the whitespace regexes, it looks like we can add a modifier to set them to binary mode: `/\s/n`:

http://ruby-doc.org/core-2.1.9/Regexp.html#class-Regexp-label-Encoding